learning_phase/chat_sample/server.go

```diff
@@ -0,0 +1,121 @@
+package main
+
+import (
+	"context"
+	"log"
+	"net"
+	"os"
+	"os/signal"
+	"strings"
+	"sync"
+)
+
+type server struct {
+	Password          string
+	ClientConnections map[string]*client
+	commands          chan command
+	mu                sync.RWMutex
+}
+
+func newServer(password string) *server {
+	return &server{
+		Password:          password,
+		ClientConnections: make(map[string]*client),
+		commands:          make(chan command),
+	}
+}
+
+func (s *server) run() {
+	// keeps receiving the commands from all clients
+	for cmd := range s.commands {
+		switch cmd.id {
+		case cmdauth:
+			s.auth(cmd.client, cmd.args[1], cmd.args[2])
+		case cmdpm:
+			s.pm(cmd.client, cmd.args)
+		case cmdmsg:
+			s.msg(cmd.client, cmd.args)
+		case cmdquit:
+			s.quit(cmd.client)
+		}
+	}
+}
+
+// when ctrl-C is pressed we want to terminate all clients
+func (s *server) gracefulexit(cancelFunc context.CancelFunc) {
+	signalChan := make(chan os.Signal)
+
+	signal.Notify(signalChan, os.Interrupt)
+
+	for {
+		sig := <-signalChan
+		switch sig {
+
+		case os.Interrupt:
+			log.Printf("SIGINT received, exiting ...")
+			cancelFunc()
```

Comment:
So you are calling the cancel function, and it's being propogated. But after you call it, you aren't really waiting for all the functions/goroutines the cancel has been propogated to to actually finish exiting, before the return statement executes. So handling this can get complicated, and you'll need to pass channel(s) to these functions, which they will write to on exiting. This function will need to wait for message(s) to be written to the channel, indicating that everything has gracefully exited, before returning. Maybe read up some more on context to get a better idea

Reply:
so I've added a finish channel that receives true when all the clients have exited and then the server will exit